runtime/browser/ui/desktop/xwalk_permission_dialog_manager.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2016 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/runtime/browser/ui/desktop/xwalk_permission_dialog_manager.h"
+
+#include "base/bind.h"
+#include "base/i18n/rtl.h"
+#include "base/strings/utf_string_conversions.h"
+#include "components/app_modal/app_modal_dialog.h"
+#include "components/app_modal/app_modal_dialog_queue.h"
+#include "components/app_modal/native_app_modal_dialog.h"
+#include "grit/components_strings.h"
+#include "xwalk/runtime/browser/ui/desktop/xwalk_permission_modal_dialog.h"
+#include "xwalk/runtime/browser/xwalk_content_settings.h"
+
+DEFINE_WEB_CONTENTS_USER_DATA_KEY(xwalk::XWalkPermissionDialogManager);
+
+namespace xwalk {
+
+XWalkPermissionDialogManager::XWalkPermissionDialogManager(
+    content::WebContents* web_contents) {
+  DCHECK(web_contents);
+}
+
+XWalkPermissionDialogManager::~XWalkPermissionDialogManager() {
+}
+
+void XWalkPermissionDialogManager::RequestPermission(
+    content::WebContents* web_contents,
+    ContentSettingsType type,
+    const GURL& origin_url,
+    const std::string& accept_lang,
+    const base::string16& message_text,
+    const base::Callback<void(bool)>& callback) {
+  ContentSetting content_setting =
+      XWalkContentSettings::GetInstance()->GetPermission(
+          type,
+          origin_url,
+          web_contents->GetLastCommittedURL().GetOrigin());
+  if (content_setting == CONTENT_SETTING_ALLOW ||
+      content_setting == CONTENT_SETTING_BLOCK) {
+    callback.Run(content_setting == CONTENT_SETTING_ALLOW ? true : false);
+    return;
+  }
+
+  app_modal::AppModalDialogQueue::GetInstance()->AddDialog(
+      new XWalkPermissionModalDialog(
+          web_contents,
+          message_text,
+          base::Bind(&XWalkPermissionDialogManager::OnPermissionDialogClosed,
+          base::Unretained(this), web_contents, type, origin_url, callback)));
```

Comment:
This will be destroyed when the WebContent is about to die. But you're right we should invalidate the dialog when this dies so that the callback will never be called.